### PR TITLE
Student academic records - k_student updates

### PR DESCRIPTION
**Pull request by sleblanc23**
## Description & motivation
`fct_student_academic_record` and its dependent models currently include `k_student` only, which excludes any historical records from years prior to the earliest `dim_student` records. This PR brings in `k_student_xyear` and updates the logic to allow `k_student` to be null.

## Changes to existing files:
- `fct_course_transcripts`, `fct_student_academic_record`, `fct_student_gpa` : Added `k_student_xyear` and `k_student` via a left join to `dim_student`; updated constraints
- yml files for these models: Updated to include a sample query for joining to the most recent `dim_student` record

## Tests and QC done (together with [edu_edfi_source):](https://github.com/edanalytics/edu_edfi_source/pull/55)
- Confirmed that row counts were unchanged in Jeffco and correct k_student was present 

## PR Merge Priority:
Low - note this depends on this [edu_edfi_source](https://github.com/edanalytics/edu_edfi_source/pull/55) PR

### CODE REVIEW

File: models/core_warehouse/fct_course_transcripts.yml [hunk 1]
```diff
@@ -22,6 +22,18 @@ models:
...
+        enrolled in 2021, then `k_student` will correspond to the 2021 `dim_student record` for all course transcript records, regardless
+        of the year of the transcript. To join to the `dim_student` record corresponding to the year of the transcript if it exists, see 
+        this sample query:
+        ```
+        SELECT *
+        FROM analytics.prod_wh.fct_course_transcripts 
+        LEFT JOIN analytics.prod_wh.dim_student
+          ON fct_course_transcripts.k_student_xyear = dim_student.k_student_xyear
+          AND fct_course_transcripts.school_year = dim_student.school_year
+        ```
```
Comment (rlittle08) [inline models/core_warehouse/fct_course_transcripts.yml:35]: Should we add a 2nd example query with join on `k_student`, and explain that you can use that to get students' demographics from latest year? Or do we want to discourage that behavior?

File: models/core_warehouse/fct_student_academic_record.yml [hunk 2]
```diff
@@ -2,5 +2,54 @@ version: 2
 
 models:
   - name: fct_student_academic_record
+    description: >
```
Comment (rlittle08) [inline models/core_warehouse/fct_student_academic_record.yml:5]: thank you for adding docs!

File: models/core_warehouse/fct_student_gpa.yml [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+        SELECT *
+        FROM analytics.prod_wh.fct_student_gpa
+        LEFT JOIN analytics.prod_wh.dim_student
+          ON fct_student_gpa.k_student_xyear = dim_student.k_student_xyear
+          AND fct_student_gpa.school_year = dim_student.school_year
+        ```
+
+    config:
+      tags: ['core']
+    columns:
```
Comment (rlittle08) [inline models/core_warehouse/fct_student_gpa.yml:28]: can you please add column descriptions for gpa_type -> max_gpa_value?

**Review by rlittle08 — CHANGES_REQUESTED** [review]

File: models/core_warehouse/fct_course_transcripts.yml (hunk 1 above)
Comment (sleblanc23) [reply models/core_warehouse/fct_course_transcripts.yml:35]: Yeah, I think that's worth doing. If nothing else it reinforces that there are multiple options that need to be considered here

File: models/core_warehouse/fct_student_gpa.sql [hunk 4]
```diff
@@ -1,13 +1,34 @@
...
 with combined_gpas as (
     select * from {{ ref('bld_ef3__combine_gpas') }}
 ),
 academic_record as (
     select * from {{ ref('fct_student_academic_record') }}
 ),
+dim_student as (
+    select * from {{ ref('dim_student') }}
+),
+most_recent_k_student as (
```
[the remark this answers is not in the document]
Comment (sleblanc23) [reply models/core_warehouse/fct_student_gpa.sql:20]: I think that's a good idea. It's also used in fct_student_parent_association, so I think that's enough use cases to add a build model for it.

I added this model and references to it. I also included `api_year` in the build model, though it's not used yet. I thought someone might want to easily reference which year of data is the most recent.

**Comment by sleblanc23** [comment]
@ejoranlienea I added k_student in the warehouse models by joining to `dim_student` on `k_student_xyear` and `school_year` rather than modifying our `gen_skey` macro. I know that goes against our typical pattern of doing all keying in stage. Do you have an opinion one way or the other?

**Review by ejoranlienea — APPROVED** [review]

**Review by ejoranlienea — APPROVED** [review]

**Review by rlittle08 — APPROVED** [review]

**Pull request merged**